crates/dht/src/time.rs

```diff
@@ -0,0 +1,1051 @@
+        factor: u8,
+        store: DynOpStore,
+    ) -> K2Result<Self> {
+        let mut pt = Self::new(origin_timestamp, factor);
+
+        pt.full_slices = store.slice_hash_count().await?;
+
+        // The end timestamp of the last full slice
+        let full_slice_end_timestamp = pt.full_slice_end_timestamp();
+
+        // Given the time reserved by full slices, how much time is left to partition into smaller slices
+        let recent_time = (Timestamp::now() - full_slice_end_timestamp).map_err(|_| {
+            K2Error::other("Failed to calculate recent time, either the clock is is wrong or this is a bug")
+        })?;
+
+        if pt.full_slices > 0 && recent_time < pt.min_recent_time {
+            return Err(K2Error::other("Not enough recent time reserved, either the clock is is wrong or this is a bug"));
+        }
+
+        // Update the state for the current time. The stored slices might be out of date.
+        pt.update(store, Timestamp::now()).await?;
+
+        Ok(pt)
+    }
+
+    /// The timestamp at which the next update is required.
+    ///
+    /// See the field [PartitionedTime::next_update_at] for more information.
+    ///
+    /// This value is updated by [PartitionedTime::update], so you can check the next update time
+    /// after calling that method.
+    pub fn next_update_at(&self) -> Timestamp {
+        self.next_update_at
+    }
+
+    /// Update the state of the [PartitionedTime] to the current time.
+    ///
+    /// This method will:
+    ///   - Check if there is space for any new full slices
+    ///   - Store the combined hash of any new full slices
+    ///   - Recompute the layout of partial slices. Each partial slice will have its hash computed
+    ///     or re-used.
+    ///   - Update the `next_update_at` field to the next time an update is required.
+    pub async fn update(
+        &mut self,
+        store: DynOpStore,
+        current_time: Timestamp,
+    ) -> K2Result<()> {
+        let mut full_slices_end_timestamp = self.full_slice_end_timestamp();
+
+        let recent_time =
+            (current_time - full_slices_end_timestamp).map_err(|_| {
+                K2Error::other(
+                    "Current time is before the complete time slice boundary",
+                )
+            })?;
+
+        // Check if there is enough time to allocate new full slices
+        let new_full_slices_count = if recent_time > self.min_recent_time {
+            // Rely on integer rounding to get the correct number of full slices
+            // that need adding.
+            (recent_time - self.min_recent_time).as_secs()
+                / ((1u64 << self.factor) * UNIT_TIME.as_secs())
+        } else {
+            0
+        };
+
+        for i in 0..new_full_slices_count {
+            // Store the hash of the full slice
+            let op_hashes = store
+                .retrieve_op_hashes_in_time_slice(
+                    full_slices_end_timestamp,
```

Comment:
It seems that this need to be updated in the loop. Otherwise it's always retrieving ops for the same time slice.

Reply:
Oh dear, that seems like something one could write a test for! That's certainly a bug. Let me get a test and a fix in for that :)

Reply:
Added a new test `update_allocate_multiple_new_complete` that was failing and now passes. Also extracted a pre-calculated full slice size because that was being done in several places. That reminded me that I'd meant to do that and fix this issue... Glad it got caught at review :)

[fffee075](https://github.com/holochain/kitsune2/pull/20/commits/fffee0753fb996628fa6e11c237dbd645392b69a)